druid-shell/src/window.rs

```diff
@@ -675,6 +686,24 @@ pub trait WinHandler {
     #[allow(unused_variables)]
     fn lost_focus(&mut self) {}
 
+    /// Called when a drop enter the window.
+    fn drop_enter(&mut self) {}
+
+    #[allow(unused_variables)]
+    fn drop_motion(&mut self, event: &DropEvent) {}
+
+    /// Called when a drop leaves the window.
+    fn drop_leave(&mut self) {}
+
+    /// Called when a drop is droped.
+    fn drop_droped(&mut self) {}
+
+    /// Called when a drag session ends
+    ///
+    /// action is None if drag got canceled
+    #[allow(unused_variables)]
+    fn drag_end(&mut self, id: DragDropId, action: Option<DragDropAction>);
```

Comment:
Various thoughts:

- even if we *initiate* a drag-drop, are still participating in it as a receiver? In this case, `drop_enter` doesn't quite make sense, since we're already going to be inside the view.
- Does `drop_motion` replace `mouse_moved`, while a drag and drop is in progress? That is, do you still receive `mouse_moved`?
- I would prefer `drop_moved` to `drop_motion`.


Okay, thinking about this a bit more: I think we can maybe unify the "initiating" and "receiving" cases.

I'm imagining something called a `DragSession`, that encapsulates all the information about a drag-and-drop operation. This is probably a replacement for `DragContext`?

This means the data, the source, the allowed operations, the thumbnails, etc.

(One complication: on macOS I'm not sure if you can start a drag explicitly? It seems to be something that you opt into, and then the OS initiates the drag when it is detected over your view?)

Anyway: my thinking is that to *start* a drag, you create one of these sessions, and submit it.

Then, when a session has begun, regardless of who initiated it, you will be called back on one of three methods:

```rust
// WindowHandle
// make sure we use the same token/id from this when we call `drag_began`
fn start_drag(&self, session: DragSession);

// WinHandler
fn drag_began(&mut self, drag: &DragSession, mouse: &MouseEvent);
fn drag_move(&mut self, drag: &DragSession, mouse: &MouseEvent);
fn drag_end(&mut self, drag: &DragSession, mouse: &MouseEvent, canceled: bool);
```

The main question I have is whether or not we can even manually begin a drag on macOS. You certainly can't begin it anytime; there needs to be a mousedown event.

It's possible that what we're going to have to do is register "drag regions", during `layout`, and then let the OS handle initiating the drag? I'm not sure if you've thought about this, but it could change the API significantly...

Also-also: I'm not sure if we need API for changing the cursor explicitly, but we probably want some way of communicating whether we're going to accept a drop, so that the platform can change the cursor as needed? This might introduce the need for a `DragContext` in addition to `DragSession`... 🤔

Reply:
> registering drag regions

I thought about it earlier but didn't know it would be *this* easy to integrate into druid. This feels like the "correct way"

> combining into DragSession

as source: we can set the allowed actions, image/icon for the drag and drop
as target: we can set current action, deny(cancel) the drag drop

I am not sure how to not let source/target use each other's methods. 
Again, gtk3 is doing this. It is confusing to say the least

> ```rust
> fn drag_began(&mut self, drag: &DragSession, mouse: &MouseEvent);
> fn drag_move(&mut self, drag: &DragSession, mouse: &MouseEvent);
> fn drag_end(&mut self, drag: &DragSession, mouse: &MouseEvent, canceled: bool);
> ```

for source, drag_began isn't useful? And most platform (except x11) don't provide move events to the source, so `drag_move` isn't called for the source. 

now `drag_end` I am not about unification of `drag_end` and `drop_dropped`  because these semantically look different to me:
as source: drag and drop finished, now delete your data, etc
as target: drag and drop finished, now add the data 


> Does drop_motion replace mouse_moved, while a drag and drop is in progress? That is, do you still receive mouse_moved?

no mouse_move during drag drop

> I would prefer drop_moved to drop_motion.

sure :+1:

Reply:
nativeshell supports starting drag drop manually on macos([code](https://github.com/nativeshell/nativeshell/blob/87031a6195fb9c6327d8f1a8d15c129adb975171/nativeshell/src/shell/platform/macos/drag_context.rs#L161))